### PR TITLE
[2d] pass shape/stride during tensor unflatten

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113547
* #113324
* #113323

as titled, built on top of the work @wz337 enabled, this could save some
runtime CPU time to recreate DTensor parameters with correct
shape/stride, and avoid issues when un-even sharding parameters

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

### CODE REVIEW

File: torch/distributed/tensor/parallel/_data_parallel_utils.py [hunk 1]
```diff
@@ -20,5 +20,7 @@ def _unflatten_tensor(tensor: torch.Tensor, spec: DTensorSpec) -> torch.Tensor:
         spec.mesh,
         spec.placements,
         run_check=False,
+        shape=spec.tensor_meta.shape,
```
Comment (XilunWu) [inline torch/distributed/tensor/parallel/_data_parallel_utils.py:23]: use `spec.shape`

File: torch/distributed/tensor/parallel/_data_parallel_utils.py [hunk 2]
```diff
@@ -20,5 +20,7 @@ def _unflatten_tensor(tensor: torch.Tensor, spec: DTensorSpec) -> torch.Tensor:
...
+        stride=spec.tensor_meta.stride,
```
Comment (XilunWu) [inline torch/distributed/tensor/parallel/_data_parallel_utils.py:24]: add `def stride` just as `shape` and `ndim`?

**Review by XilunWu — COMMENTED** [review]
LGTM! Ready to accept when these 2 comments get addressed.

**Review by XilunWu — APPROVED** [review]
LGTM!

**Review by wz337 — APPROVED** [review]
LGTM!

**Comment by wanchaol** [comment]
this relanded as a separate PR, closing this

**Pull request closed without merging**